File: app/src/main/java/org/codepond/commitbrowser/common/ui/ViewModelFactory.java
package org.codepond.commitbrowser.common.ui;

import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;

import org.codepond.commitbrowser.api.GithubApi;
import org.codepond.commitbrowser.commitdetail.CommitDetailViewModel;
import org.codepond.commitbrowser.commitlist.CommitListViewModel;

import javax.inject.Inject;

public class ViewModelFactory implements ViewModelProvider.Factory {
    private GithubApi githubApi;

    @Inject
    public ViewModelFactory(GithubApi githubApi) {
        this.githubApi = githubApi;
    }

    @Override
    public <T extends ViewModel> T create(Class<T> modelClass) {
        if (modelClass.isAssignableFrom(CommitDetailViewModel.class)) {
            return (T) new CommitDetailViewModel(githubApi);
        }
        else if (modelClass.isAssignableFrom(CommitListViewModel.class)) {
            return (T) new CommitListViewModel(githubApi);
        }
        throw new IllegalArgumentException("Unknown ViewModel class");
    }
}

File: app/src/main/java/org/codepond/commitbrowser/common/ui/ViewModelFactory.kt
package org.codepond.commitbrowser.common.ui

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider

import org.codepond.commitbrowser.api.GithubApi
import org.codepond.commitbrowser.commitdetail.CommitDetailViewModel
import org.codepond.commitbrowser.commitlist.CommitListViewModel

import javax.inject.Inject

class ViewModelFactory @Inject
constructor(private val githubApi: GithubApi) : ViewModelProvider.Factory {

    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(CommitDetailViewModel::class.java)) {
            return CommitDetailViewModel(githubApi) as T
        } else if (modelClass.isAssignableFrom(CommitListViewModel::class.java)) {
            return CommitListViewModel(githubApi) as T
        }
        throw IllegalArgumentException("Unknown ViewModel class")
    }
}
